odo/backends/url.py

```diff
@@ -0,0 +1,129 @@
+from __future__ import print_function, division, absolute_import
+
+import os
+import uuid
+import zlib
+import re
+
+from contextlib import contextmanager
+
+import requests
+from toolz import memoize
+
+from .. import (discover, CSV, resource, append, convert, drop, Temp, JSON,
+                JSONLines, chunks)
+
+from multipledispatch import MDNotImplementedError
+
+from .text import TextFile
+
+from ..compatibility import urlparse
+from ..utils import tmpfile, ext, sample, filter_kwargs
+
+
+class _URL(object):
+    """ Parent class for data accessed through ``URLs``
+
+    Parameters
+    ----------
+
+    url : str
+        full url to file
+    chunk_size : int (default 1024)
+        Size of chunks streamed into memory
+    decode_unicode : bool (default False)
+        If True, content will be decoded using the best available encoding based on the response.
+
+     Examples
+    --------
+
+    >>> from odo import URL, CSV
+    >>> u = URL(CSV)('http://foo.com/myfile.csv')
+
+    Normally create through resource uris
+
+    >>> data = resource('http://foo.com/myfile.csv')
+    >>> data.url
+    'http://foo.com/myfile.csv'
+    >>> data.filename
+    'myfile.csv'
+    """
+    def __init__(self, url, chunk_size=1024, decode_unicode=False, *args,
+                 **kwargs):
+
+        self.subtype.__init__(self, url, *args, **kwargs)
+
+        self.url = url
+        self.chunk_size = chunk_size
+        self.decode_unicode = decode_unicode
+        self.filename = os.path.basename(urlparse(url).path)
+
+
+def URL(cls):
+    return type('URL(%s)' % cls.__name__, (_URL, cls), {'subtype': cls})
+
+URL.__doc__ = _URL.__doc__
+URL = memoize(URL) # what's happening here
```

Comment:
Yes, exactly!  This threw me for a bit.  Before I had the `memoize` line i was seeing things like:

```
In [3]: isinstance(data, URL(CSV))
Out[3]: False

In [4]: data
Out[4]: <odo.backends.url.URL(CSV) at 0x109a4bb90>

In [5]: URL(CSV)
Out[5]: odo.backends.url.URL(CSV)
```